### PR TITLE
python3Packages.warp-lang: init at 1.7.2.post1

### PR DESCRIPTION
**Pull request by yzx9**
- Code: https://github.com/NVIDIA/warp
- Homepage: https://nvidia.github.io/warp/
- Release: https://github.com/NVIDIA/warp/releases/tag/v1.7.2.post1

Warp is a Python framework for writing high-performance simulation and graphics code. Warp takes regular Python functions and JIT compiles them to efficient kernel code that can run on the CPU or GPU.

Warp is designed for spatial computing and comes with a rich set of primitives that make it easy to write programs for physics simulation, perception, robotics, and geometry processing. In addition, Warp kernels are differentiable and can be used as part of machine-learning pipelines with frameworks such as PyTorch, JAX and Paddle.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by yzx9** [comment]
Oh, sorry. I didn’t realize that adding the cuda label would automatically add it to the project.

File: pkgs/development/python-modules/warp-lang/default.nix [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+      #
+      # See: https://github.com/NVIDIA/warp/issues/552
+      substituteInPlace warp/tests/test_fem.py \
+        --replace-fail "add_function_test(TestFem, \"test_integrate_gradient\", test_integrate_gradient, devices=devices)" ""
+    '';
+
+  build-system = [
+    build
+    setuptools
+    wheel
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/warp-lang/default.nix]:
```suggestion
```
This is not needed.

File: pkgs/development/python-modules/warp-lang/default.nix [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    ''
+    + lib.optionalString stdenv.hostPlatform.isAarch64 ''
+      # Broken test on aarch64. Since unittest doesn't support disabling
+      # a single test, and pytest isn't compatible, we patch the test
+      # file directly instead.
+      #
+      # See: https://github.com/NVIDIA/warp/issues/552
+      substituteInPlace warp/tests/test_fem.py \
+        --replace-fail "add_function_test(TestFem, \"test_integrate_gradient\", test_integrate_gradient, devices=devices)" ""
+    '';
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/warp-lang/default.nix]:
```suggestion
    # Broken test on aarch64. Since unittest doesn't support disabling
    # a single test, and pytest isn't compatible, we patch the test
    # file directly instead.
    #
    # See: https://github.com/NVIDIA/warp/issues/552
    + lib.optionalString stdenv.hostPlatform.isAarch64 ''
      substituteInPlace warp/tests/test_fem.py \
        --replace-fail "add_function_test(TestFem, \"test_integrate_gradient\", test_integrate_gradient, devices=devices)" ""
    '';
```
Nit: it is preferable to keep comments out of Nix strings. It allows them to be changed without causing rebuilds.

**Review by GaetanLepage — COMMENTED** [review]
Are you able to build the package? It fails on my system.

Very minor feedback:

File: pkgs/development/python-modules/warp-lang/default.nix (hunk 2 above)
Comment (yzx9) [reply pkgs/development/python-modules/warp-lang/default.nix]: Oh, that's a good reason to keep the comments in the Nix expression, thanks!

**Comment by yzx9** [comment]
@GaetanLepage Thanks for your review! I was able to build the package successfully on both `aarch64-darwin` and `x86_64-linux` ([OfBorg can handle that too](https://github.com/NixOS/nixpkgs/runs/43484087195)). On Linux, I tested it both with and without `cudaSupport` and `libmathdSupport`.

I did notice that the `aarch64-linux` build failed on OfBorg due to [a segmentation fault during unit tests](https://github.com/NixOS/nixpkgs/runs/43595848844). I suspect it’s related to the same SIM module issue I mentioned earlier, and I’m currently working on reproducing it.

**Update**: ~~I wasn’t able to reproduce the issue on my `aarch64-linux` environment, which is emulated on `x86_64-linux`.~~ No, I can

<details>
<summary>Update 2: I can reproduce the build failure after rebase to latest commits, caused by paddlepaddle interop tests</summary>

```console
test_dlpack_warp_to_torch_cpu (warp.tests.interop.test_dlpack.TestDLPack.test_dlpack_warp_to_torch_cpu) ... ok
test_dlpack_warp_to_torch_v2_cpu (warp.tests.interop.test_dlpack.TestDLPack.test_dlpack_warp_to_torch_v2_cpu) ... ok
test_dlpack_warp_to_warp_cpu (warp.tests.interop.test_dlpack.TestDLPack.test_dlpack_warp_to_warp_cpu) ... ok
test_device_conversion_cpu (warp.tests.interop.test_jax.TestJax.test_device_conversion_cpu) ... ok
test_dtype_from_jax (warp.tests.interop.test_jax.TestJax.test_dtype_from_jax) ... ok
test_dtype_to_jax (warp.tests.interop.test_jax.TestJax.test_dtype_to_jax) ... ok
test_array_ctype_from_paddle_cpu (warp.tests.interop.test_paddle.TestPaddle.test_array_ctype_from_paddle_cpu) ... FAIL
test_device_conversion_cpu (warp.tests.interop.test_paddle.TestPaddle.test_device_conversion_cpu) ... ok
test_direct_cpu (warp.tests.interop.test_paddle.TestPaddle.test_direct_cpu) ... malloc(): unaligned tcache chunk detected
/nix/store/c15sc6xcm41dd11draflv8nn84br7zki-unittest-check-hook/nix-support/setup-hook: line 6:  1324 Aborted                 (core dumped) /nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/bin/python3.12 -m unittest discover "${flagsArray[@]}"
```
</details>

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 412838`
Commit: `fd636882374cdce4c1e042b088b4eeaa18bbdb69`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python312Packages.warp-lang</li>
    <li>python312Packages.warp-lang.dist</li>
    <li>python313Packages.warp-lang</li>
    <li>python313Packages.warp-lang.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python312Packages.warp-lang</li>
    <li>python312Packages.warp-lang.dist</li>
    <li>python313Packages.warp-lang</li>
    <li>python313Packages.warp-lang.dist</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python312Packages.warp-lang</li>
    <li>python312Packages.warp-lang.dist</li>
    <li>python313Packages.warp-lang</li>
    <li>python313Packages.warp-lang.dist</li>
  </ul>
</details>

**Review by GaetanLepage — APPROVED** [review]
LGTM, thanks!

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 412838`
Commit: `fd636882374cdce4c1e042b088b4eeaa18bbdb69`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python312Packages.warp-lang</li>
    <li>python312Packages.warp-lang.dist</li>
    <li>python313Packages.warp-lang</li>
    <li>python313Packages.warp-lang.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python312Packages.warp-lang</li>
    <li>python312Packages.warp-lang.dist</li>
    <li>python313Packages.warp-lang</li>
    <li>python313Packages.warp-lang.dist</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python312Packages.warp-lang</li>
    <li>python312Packages.warp-lang.dist</li>
    <li>python313Packages.warp-lang</li>
    <li>python313Packages.warp-lang.dist</li>
  </ul>
</details>

**Comment by yzx9** [comment]
@GaetanLepage Thank you!

File: pkgs/development/python-modules/warp-lang/default.nix [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+          nvidiaCudaRedist
+          // {
+            url = "https://developer.download.nvidia.cn/compute/mathdx/License.txt";
+          }
+        )
+
+        # Some of the libmathdx routines were written by or derived
+        # from code written by Meta Platforms, Inc. and affiliates and
+        # are subject to the BSD License.
+        bsd
```
Comment (mweinelt) [inline pkgs/development/python-modules/warp-lang/default.nix:100]:
```
nix-repl> lib.licenses.bsd
error:
       … while evaluating the attribute 'licenses.bsd'
         at /nixos/master/lib/default.nix:76:7:
           75|       # constants
           76|       licenses = callLibs ./licenses.nix;
             |       ^
           77|       sourceTypes = callLibs ./source-types.nix;

       error: attribute 'bsd' missing
       at «string»:1:14:
            1| lib.licenses.bsd
             |              ^
       Did you mean one of bsd0, bsd1, bsd2, bsd3 or amd?
```

That is not a valid license specifier.

Comment (ConnorBaker) [reply pkgs/development/python-modules/warp-lang/default.nix:100]: https://github.com/NixOS/nixpkgs/pull/419722

Comment (GaetanLepage) [reply pkgs/development/python-modules/warp-lang/default.nix:100]: Thanks
